server/sonar-server/src/main/java/org/sonar/server/computation/step/PersistComponentsStep.java

```diff
@@ -0,0 +1,244 @@
+    component.setEnabled(true);
+    component.setScope(getScope(reportComponent));
+    component.setQualifier(getQualifier(reportComponent));
+    component.setName(getFileName(reportComponent));
+
+    if (isProjectOrModule(reportComponent)) {
+      component.setLongName(component.name());
+      if (reportComponent.hasDescription()) {
+        component.setDescription(reportComponent.getDescription());
+      }
+    } else {
+      component.setLongName(reportComponent.getPath());
+      if (reportComponent.hasPath()) {
+        component.setPath(reportComponent.getPath());
+      }
+      if (reportComponent.hasLanguage()) {
+        component.setLanguage(reportComponent.getLanguage());
+      }
+    }
+    if (parentModule != null) {
+      component.setParentProjectId(parentModule.getId());
+      component.setProjectUuid(parentModule.projectUuid());
+      component.setModuleUuid(parentModule.uuid());
+      component.setModuleUuidPath(reportComponent.getType().equals(Constants.ComponentType.MODULE) ?
+        parentModule.moduleUuidPath() + component.uuid() + ComponentDto.MODULE_UUID_PATH_SEP :
+        parentModule.moduleUuidPath());
+    } else {
+      component.setProjectUuid(uuid);
+      component.setModuleUuidPath(ComponentDto.MODULE_UUID_PATH_SEP + component.uuid() + ComponentDto.MODULE_UUID_PATH_SEP);
+    }
+    return component;
+  }
+
+  private boolean updateComponent(ComponentDto existingComponent, ComponentDto newComponent) {
+    boolean isUpdated = false;
+    if (Scopes.PROJECT.equals(existingComponent.scope())) {
+      if (!newComponent.name().equals(existingComponent.name())) {
+        existingComponent.setName(newComponent.name());
+        isUpdated = true;
+      }
+      if (!StringUtils.equals(existingComponent.description(), newComponent.description())) {
+        existingComponent.setDescription(newComponent.description());
+        isUpdated = true;
+      }
+    }
+
+    if (!StringUtils.equals(existingComponent.moduleUuid(), newComponent.moduleUuid())) {
+      existingComponent.setModuleUuid(newComponent.moduleUuid());
+      isUpdated = true;
+    }
+    if (!existingComponent.moduleUuidPath().equals(newComponent.moduleUuidPath())) {
+      existingComponent.setModuleUuidPath(newComponent.moduleUuidPath());
+      isUpdated = true;
+    }
+    if (!ObjectUtils.equals(existingComponent.parentProjectId(), newComponent.parentProjectId())) {
+      existingComponent.setParentProjectId(newComponent.parentProjectId());
+      isUpdated = true;
+    }
+
+    return isUpdated;
+  }
+
+  private static boolean isProjectOrModule(BatchReport.Component reportComponent) {
+    return reportComponent.getType().equals(Constants.ComponentType.PROJECT) || reportComponent.getType().equals(Constants.ComponentType.MODULE);
+  }
+
+  private static String getScope(BatchReport.Component reportComponent) {
+    switch (reportComponent.getType()) {
+      case PROJECT:
+        return Scopes.PROJECT;
+      case MODULE:
```

Comment:
![New](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/new.png) ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Either merge this case with the identical one on line "182" or change one of the implementations. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1871)